### PR TITLE
feat: Partition memtables by time if compaction window is provided

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR introduces a new struct `TimePartitions` to partition mutable memtables by time.
- It uses the compaction time window (`compaction.twcs.time_window`) as the partition time range
- Each partition holds timestamps between `[start_time, start_time + time_window)`
- We uses the `timestamp / time_window * time_window` to locate the proper partition
- If the compaction time window is not provided, it only has one partition and all timestamps belong to that partition
- Once we freeze a memtable version,  we move all memtables in `TimePartitions` to immutable memtable list.

With this feature, a region can strictly divide data by the time in memory. It also ensures flushed SSTs only belong to a single time window.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/memtable/time_partition.rs [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    fn write_multi_parts(&self, kvs: &KeyValues, parts: &PartitionVec) -> Result<()> {
+        let mut parts_to_write = HashMap::new();
+        let mut missing_parts = HashMap::new();
+        for kv in kvs.iter() {
+            let mut part_found = false;
+            // Safety: We used the timestamp before.
+            let ts = kv.timestamp().as_timestamp().unwrap().unwrap();
+            for part in parts {
+                if part.contains_timestamp(ts) {
+                    // Safety: If time range is None then we won't go to this method.
```
Comment (MichaelScofield) [inline src/mito2/src/memtable/time_partition.rs:243]: But in `contains_timestamp` method, if `time_range` is None, it returns true: 
```rust
let Some(range) = self.time_range else {
    return true;
};
```

File: src/mito2/src/memtable/time_partition.rs [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+
+        Ok(())
+    }
+}
+
+/// Computes the start timestamp of the partition for `ts`.
+///
+/// It always use bucket size in seconds which should fit all timestamp resolution.
+fn partition_start_timestamp(ts: Timestamp, bucket: Duration) -> Option<Timestamp> {
+    // Safety: We convert it to seconds so it never returns Some.
```
Comment (MichaelScofield) [inline src/mito2/src/memtable/time_partition.rs:337]: // Safety: We convert it to seconds so it never returns ~~Some~~ None. ?

File: src/mito2/src/memtable/time_partition.rs (hunk 1 above)
Comment (evenyag) [reply src/mito2/src/memtable/time_partition.rs:243]: There is an invariant: If part_duration is None, then there will be only one partition and the time range of the partition is None; If part_duration is Some, then all partitions' time ranges are always Some.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**